File: Meta/809.expressive-words.java
/**
 * Algorithm:
 * Compress the target and each word into character groups. A word group must
 * have the same character, no larger count than the target group, and if the
 * target group length is below 3 then the counts must match exactly.
 *
 * Java data structures:
 * Arrays of small Group objects keep the comparison explicit.
 *
 * Complexity:
 * Time O(|s| + total word length), space O(|s|) for target groups and O(word)
 * temporary groups.
 */
class Solution {
    public int expressiveWords(String s, String[] words) {
        Group[] target = groups(s);
        int ans = 0;
        for (String word : words) {
            Group[] current = groups(word);
            if (current.length != target.length) {
                continue;
            }
            boolean ok = true;
            for (int i = 0; i < target.length; i++) {
                int tn = target[i].count;
                int wn = current[i].count;
                if (target[i].ch != current[i].ch || wn > tn || (tn < 3 && wn != tn)) {
                    ok = false;
                    break;
                }
            }
            if (ok) {
                ans++;
            }
        }
        return ans;
    }

    private Group[] groups(String t) {
        java.util.List<Group> res = new java.util.ArrayList<>();
        int i = 0;
        while (i < t.length()) {
            int j = i + 1;
            while (j < t.length() && t.charAt(j) == t.charAt(i)) {
                j++;
            }
            res.add(new Group(t.charAt(i), j - i));
            i = j;
        }
        return res.toArray(new Group[0]);
    }

    private static class Group {
        char ch;
        int count;

        Group(char ch, int count) {
            this.ch = ch;
            this.count = count;
        }
    }
}

File: Amazon/809.expressive-words.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
    vector<pair<char, int>> groups(const string& s) {
        vector<pair<char, int>> res;
        for (int i = 0; i < (int)s.size();) {
            int j = i + 1;
            while (j < (int)s.size() && s[j] == s[i]) ++j;
            res.push_back({s[i], j - i});
            i = j;
        }
        return res;
    }

public:
    int expressiveWords(string s, vector<string>& words) {
        auto target = groups(s);
        int ans = 0;
        for (const string& word : words) {
            auto g = groups(word);
            if (g.size() != target.size()) continue;
            bool ok = true;
            for (int i = 0; i < (int)g.size(); ++i) {
                if (g[i].first != target[i].first || g[i].second > target[i].second || (target[i].second < 3 && g[i].second != target[i].second)) {
                    ok = false;
                    break;
                }
            }
            ans += ok;
        }
        return ans;
    }
};

/*
Interview explanation:
Run-length encode s and each word. Characters must match group by group; a word group may be shorter only when the target group length is at least three.

C++ data structures: vector<pair<char,int>> stores compact character runs.

Edge cases: target runs of length 1 or 2 cannot be stretched, so counts must match exactly.

Complexity: O(total characters) time and O(number of groups) space per processed string.
*/
